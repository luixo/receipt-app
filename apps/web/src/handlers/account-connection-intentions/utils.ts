import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
