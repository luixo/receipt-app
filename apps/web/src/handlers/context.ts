import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
