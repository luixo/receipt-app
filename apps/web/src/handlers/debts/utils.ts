import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
