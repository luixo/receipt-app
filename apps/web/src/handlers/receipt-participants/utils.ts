import type { Selection } from "kysely";

import type { ReceiptsId, UsersId } from "~db/models";
import type {
	Database,
	ReceiptsDatabase,
	ReceiptsSelectExpression,
} from "~db/types";

export const getReceiptParticipant = <
	SE extends ReceiptsSelectExpression<"receiptParticipants">,
>(
	database: Database,
	userId: UsersId,
	receiptId: ReceiptsId,
	selectExpression: SE[],
): Promise<
	Selection<ReceiptsDatabase, "receiptParticipants", SE> | undefined
> =>
	database
		.selectFrom("receiptParticipants")
		.where((eb) => eb.and({ receiptId, userId }))
		.select(selectExpression)
		.limit(1)
		.executeTakeFirst();
