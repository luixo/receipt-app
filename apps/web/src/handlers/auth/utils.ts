import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
