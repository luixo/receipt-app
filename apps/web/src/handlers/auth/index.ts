import { procedure as confirmEmail } from "./confirm-email";
import { procedure as login } from "./login";
import { procedure as register } from "./register";
import { procedure as resetPassword } from "./reset-password";
import { procedure as voidAccount } from "./void-account";

export const router = {
	login,
	register,
	resetPassword,
	confirmEmail,
	voidAccount,
};
