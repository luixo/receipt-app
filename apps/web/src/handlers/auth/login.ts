import { TRPCError } from "@trpc/server";
import { z } from "zod";

import { passwordSchema } from "~app/utils/validation";
import { createAuthorizationSession } from "~web/handlers/auth/utils";
import { unauthProcedure } from "~web/handlers/trpc";
import { emailSchema } from "~web/handlers/validation";
import { getHash } from "~web/utils/crypto";
import { setAuthCookie } from "~web/utils/server-cookies";

export const procedure = unauthProcedure
	.input(
		z.strictObject({
			email: emailSchema,
			password: passwordSchema,
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const { database } = ctx;
		const result = await database
			.selectFrom("accounts")
			.where("email", "=", input.email.lowercase)
			.innerJoin("users", (qb) =>
				qb.onRef("users.connectedAccountId", "=", "accounts.id"),
			)
			.whereRef("users.id", "=", "users.connectedAccountId")
			.select([
				"accounts.id as accountId",
				"accounts.email",
				"accounts.passwordSalt",
				"accounts.passwordHash",
				"users.name",
				"accounts.confirmationToken",
				"accounts.avatarUrl",
			])
			.executeTakeFirst();

		if (!result) {
			const errorMessage = `Authentication of account "${input.email.original}" failed: account not found.`;
			ctx.logger.debug(errorMessage);
			throw new TRPCError({
				code: "UNAUTHORIZED",
				message: errorMessage,
			});
		}
		const isPasswordValid =
			getHash(input.password, result.passwordSalt) === result.passwordHash;
		if (!isPasswordValid) {
			const errorMessage = `Authentication of account "${input.email.original}" failed: password is wrong.`;
			ctx.logger.debug(errorMessage);
			throw new TRPCError({
				code: "UNAUTHORIZED",
				message: errorMessage,
			});
		}
		const { authToken, expirationDate } = await createAuthorizationSession(
			ctx,
			result.accountId,
		);
		ctx.logger.debug(
			`Authentication of account "${input.email.original}" succeed.`,
		);
		setAuthCookie(ctx.res, authToken, expirationDate);
		return {
			account: {
				id: result.accountId,
				verified: !result.confirmationToken,
				avatarUrl: result.avatarUrl || undefined,
			},
			user: {
				name: result.name,
			},
		};
	});
