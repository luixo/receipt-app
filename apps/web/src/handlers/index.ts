import { t } from "~web/handlers/trpc";

import { router as accountRouter } from "./account/index";
import { router as accountConnectionIntentionsRouter } from "./account-connection-intentions/index";
import { router as accountSettingsRouter } from "./account-settings/index";
import { router as adminRouter } from "./admin/index";
import { router as authRouter } from "./auth/index";
import { router as currencyRouter } from "./currency/index";
import { router as debtIntentionsRouter } from "./debt-intentions/index";
import { router as debtsRouter } from "./debts/index";
import { router as receiptItemConsumersRouter } from "./receipt-item-consumers/index";
import { router as receiptItemPayersRouter } from "./receipt-item-payers/index";
import { router as receiptItemsRouter } from "./receipt-items/index";
import { router as receiptParticipantsRouter } from "./receipt-participants/index";
import { router as receiptsRouter } from "./receipts/index";
import { router as resetPasswordIntentionsRouter } from "./reset-password-intentions/index";
import { router as sessionsRouter } from "./sessions/index";
import { router as usersRouter } from "./users/index";
import { router as utilsRouter } from "./utils/index";

export const router = t.router({
	// No auth
	sessions: sessionsRouter,
	auth: authRouter,
	resetPasswordIntentions: resetPasswordIntentionsRouter,
	utils: utilsRouter,
	// Auth
	account: accountRouter,
	accountSettings: accountSettingsRouter,
	receipts: receiptsRouter,
	receiptItems: receiptItemsRouter,
	users: usersRouter,
	receiptParticipants: receiptParticipantsRouter,
	currency: currencyRouter,
	receiptItemConsumers: receiptItemConsumersRouter,
	receiptItemPayers: receiptItemPayersRouter,
	accountConnectionIntentions: accountConnectionIntentionsRouter,
	debts: debtsRouter,
	debtIntentions: debtIntentionsRouter,
	// Admin
	admin: adminRouter,
});
