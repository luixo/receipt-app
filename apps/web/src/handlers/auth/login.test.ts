import { faker } from "@faker-js/faker";
import { assert, describe, expect } from "vitest";

import {
	MAX_PASSWORD_LENGTH,
	MIN_PASSWORD_LENGTH,
} from "~app/utils/validation";
import { createContext } from "~tests/backend/utils/context";
import { insertAccountWithSession } from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";
import { getResHeaders } from "~web/utils/headers";

import { procedure } from "./login";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("auth.login", () => {
	describe("input verification", () => {
		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const caller = createCaller(await createContext(ctx));
				await expectTRPCError(
					() =>
						caller.procedure({
							email: "invalid@@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		describe("password", () => {
			test("minimal length", async ({ ctx }) => {
				const caller = createCaller(await createContext(ctx));
				await expectTRPCError(
					() =>
						caller.procedure({
							email: "valid@mail.org",
							password: "a".repeat(MIN_PASSWORD_LENGTH - 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Minimal length for password is ${MIN_PASSWORD_LENGTH}`,
				);
			});

			test("maximum length", async ({ ctx }) => {
				const caller = createCaller(await createContext(ctx));
				await expectTRPCError(
					() =>
						caller.procedure({
							email: "valid@mail.org",
							password: "a".repeat(MAX_PASSWORD_LENGTH + 1),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "password": Maximum length for password is ${MAX_PASSWORD_LENGTH}`,
				);
			});
		});

		test("account not found", async ({ ctx }) => {
			const caller = createCaller(await createContext(ctx));
			const email = faker.internet.email();
			await expectTRPCError(
				() =>
					caller.procedure({
						email,
						password: "a".repeat(MIN_PASSWORD_LENGTH),
					}),
				"UNAUTHORIZED",
				`Authentication of account "${email}" failed: account not found.`,
			);
		});

		test("authentication failed", async ({ ctx }) => {
			const {
				account: { email, password },
			} = await insertAccountWithSession(ctx);
			const caller = createCaller(await createContext(ctx));
			await expectTRPCError(
				() =>
					caller.procedure({
						email,
						password: `${password}_fail`,
					}),
				"UNAUTHORIZED",
				`Authentication of account "${email}" failed: password is wrong.`,
			);
		});
	});

	describe("functionality", () => {
		test("login successful", async ({ ctx }) => {
			const {
				accountId,
				account: { email, password, avatarUrl },
				name,
			} = await insertAccountWithSession(ctx);
			const context = await createContext(ctx);
			const caller = createCaller(context);
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({ email, password }),
			);
			expect(result).toEqual<typeof result>({
				account: { id: accountId, verified: true, avatarUrl, role: undefined },
				user: { name },
			});
			const responseHeaders = getResHeaders(context.res);
			const setCookieTuple = responseHeaders.find(
				([key]) => key === "set-cookie",
			);
			assert(
				setCookieTuple,
				"Header 'set-cookie' has to be set in the response",
			);
			const tokenMatch = /authToken=([^;]+)/.exec(setCookieTuple[1].toString());
			assert(tokenMatch, "Cookie 'authToken' should be present");
			const token = tokenMatch[1];
			expect(responseHeaders).toStrictEqual<typeof responseHeaders>([
				[
					"set-cookie",
					`authToken=${token}; Path=/; Expires=Fri, 31 Jan 2020 00:00:00 GMT; HttpOnly; SameSite=Strict`,
				],
			]);
		});

		test("login successful - unverified user", async ({ ctx }) => {
			const {
				accountId,
				account: { email, password },
				name,
			} = await insertAccountWithSession(ctx, {
				account: { confirmation: {}, avatarUrl: null },
			});
			const context = await createContext(ctx);
			const caller = createCaller(context);
			const result = await caller.procedure({ email, password });
			expect(result).toEqual<typeof result>({
				account: {
					id: accountId,
					verified: false,
					avatarUrl: undefined,
					role: undefined,
				},
				user: { name },
			});
		});

		test("login successful - with different casing", async ({ ctx }) => {
			const {
				account: { email, password },
			} = await insertAccountWithSession(ctx);
			const context = await createContext(ctx);
			const caller = createCaller(context);
			await caller.procedure({ email: email.toUpperCase(), password });
		});
	});
});
