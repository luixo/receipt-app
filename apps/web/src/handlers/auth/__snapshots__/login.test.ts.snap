// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`auth.login > functionality > login successful 1`] = `
{
  "added": {
    "sessions": {
      "sessionId:3a5a4d2e-cde5-4132-9be0-c13d55fb2b6c": {
        "accountId": "5acba26a-7c4b-4340-9aa6-0a1942365f2e",
        "createdAt": "2020-01-01T00:00:00.000Z",
        "expirationTimestamp": "2020-01-31T00:00:00.000Z",
        "sessionId": "3a5a4d2e-cde5-4132-9be0-c13d55fb2b6c",
      },
    },
  },
}
`;
